presto-main-base/src/main/java/com/facebook/presto/sql/planner/optimizations/SortedExchangeRule.java

```diff
@@ -0,0 +1,196 @@
+         * This is beneficial for distributed queries where we can sort during the shuffle operation instead of
+         * adding an explicit SortNode.
+         *
+         * @param plan The plan node that needs sorting
+         * @param orderingScheme The required ordering scheme
+         * @return Optional containing the enhanced exchange node if push-down is possible, empty otherwise
+         */
+        private Optional<PlanNode> pushSortToExchangeIfPossible(PlanNode plan, OrderingScheme orderingScheme)
+        {
+            // Check if this is a suitable exchange node for sort push-down
+            if (!(plan instanceof ExchangeNode)) {
+                log.debug("Unable to push sorting to exchange because child is not exchange node");
+                return Optional.empty();
+            }
+
+            ExchangeNode exchangeNode = (ExchangeNode) plan;
+
+            // Only push sort down to REPARTITION exchanges in remote scope
+            // These are the exchanges that involve shuffling data between executors
+            if (exchangeNode.getType() != ExchangeNode.Type.REPARTITION ||
+                    !exchangeNode.getScope().isRemote()) {
+                log.debug("Unable to push sorting to exchange because it is not REMOTE REPARTITION exchange " +
+                        "(Type: {}, Scope: {})", exchangeNode.getType(), exchangeNode.getScope());
+                return Optional.empty();
+            }
+
+            // Don't push down if the exchange already has ordering requirements
+            if (exchangeNode.getOrderingScheme().isPresent()) {
+                log.debug("Unable to push sorting to exchange because exchange already has ordering scheme");
+                return Optional.empty();
+            }
+
+            // Check if all variables in the ordering scheme are available in the exchange output
+            if (!exchangeNode.getOutputVariables().containsAll(orderingScheme.getOrderByVariables())) {
+                log.debug("Unable to push sorting to exchange because not all ordering variables are available in exchange output");
+                return Optional.empty();
+            }
+
+            // Create a new sorted exchange node
+            try {
+                ExchangeNode sortedExchange = ExchangeNode.sortedPartitionedExchange(
+                        idAllocator.getNextId(),
+                        exchangeNode.getScope(),
+                        exchangeNode.getSources().get(0),
+                        exchangeNode.getPartitioningScheme().getPartitioning(),
+                        exchangeNode.getPartitioningScheme().getHashColumn(),
+                        orderingScheme);
+
+                log.info("Successfully pushed sorting down to REMOTE REPARTITION exchange with orderingScheme=%s",
+                        orderingScheme);
+                return Optional.of(sortedExchange);
+            }
+            catch (Exception e) {
```

Comment:
Curious what kinds of exception can happen here? We pass some `null` to the constructor?